common/namer/namer.go

```diff
@@ -1,57 +1,132 @@
-	childNamer := Namer{}
-	if n.prefix != "" {
-		childNamer.prefix = n.prefix + nameSep
+	return Namer{
+		ctx:      n.ctx,
+		prefixes: append(n.prefixes, prefix),
 	}
-	childNamer.prefix += prefix
-	return childNamer
 }
 
 func (n Namer) ResourceName(parts ...string) string {
 	if len(parts) == 0 {
 		panic("Resource name requires at least one part to generate name")
 	}
 
-	var resourceName string
-	if n.prefix != "" {
-		resourceName += n.prefix + nameSep
-	}
-
-	return resourceName + strings.Join(parts, nameSep)
+	return joinWithMaxLength(math.MaxInt, append(n.prefixes, parts...))
 }
 
-func (n Namer) DisplayName(parts ...pulumi.StringInput) pulumi.StringInput {
+func (n Namer) DisplayName(maxLen int, parts ...pulumi.StringInput) pulumi.StringInput {
 	var convertedParts []interface{}
 	for _, part := range parts {
 		convertedParts = append(convertedParts, part)
 	}
 	return pulumi.All(convertedParts...).ApplyT(func(args []interface{}) string {
-		strArgs := []string{n.ctx.Stack()}
+		strArgs := make([]string, 1, 1+len(n.prefixes)+len(args))
+		strArgs[0] = n.ctx.Stack()
+		strArgs = append(strArgs, n.prefixes...)
 		for _, arg := range args {
 			strArgs = append(strArgs, arg.(string))
 		}
-		return n.ResourceName(strArgs...)
+		return joinWithMaxLength(maxLen, strArgs)
 	}).(pulumi.StringOutput)
 }
+
+func joinWithMaxLength(maxLength int, tokens []string) string {
+	// Check if non-truncated concatenation fits inside maximum length
+	if lo.Sum(lo.Map(tokens, func(s string, _ int) int { return len(s) }))+(len(tokens)-1)*len(nameSep) <= maxLength {
+		return strings.Join(tokens, nameSep)
+	}
+
+	// If a truncation is needed, a hash will be needed
+	hasher := fnv.New64a()
+	for _, tok := range tokens {
+		_, _ = io.WriteString(hasher, tok)
+	}
+	fullhash := fmt.Sprintf("%016x", hasher.Sum64())
+
+	// Compute the size of the hash suffix that will be appended to the output
+	hash := fullhash
+	hashSize := maxLength * hashSizePercent / 100
+	if len(hash) > hashSize {
+		hash = hash[:hashSize]
+	} else {
+		hashSize = len(hash)
+	}
+
+	// For test purpose, we have an option to completely strip the output of the hash suffix
+	// At this point, `hashSize` is the size of the hash suffix without the dash separator
+	// -1 means that we want to also strip the dash
+	if noHash {
+		hashSize = -1
+	}
+
+	// If there’s so many tokens that truncating all of them to a single character string and keeping only the dash separators
+	// would exceed the maximum length, we cannot do anything better than returning only the hash.
+	if len(tokens)+(len(tokens))*len(nameSep)+hashSize > maxLength {
+		if len(fullhash) > maxLength {
+			return fullhash[:maxLength]
+		}
+		return fullhash
+	}
+
+	var sb strings.Builder
+
+	// Truncate all tokens in the same relative proportion
+	totalOutputSize := maxLength - len(tokens)*len(nameSep) - hashSize
+	totalInputSize := lo.Sum(lo.Map(tokens, func(s string, _ int) int { return len(s) }))
+	prevY := 0
```

Comment:
The goal is to truncate all the tokens with the same ratio.
If the token lengths could be rational numbers, the formula would have just been:
```go
ratio := totalOutputSize / totalInputSize  // At this point, we have the guarantee that totalOutputSize < totalInputSize
tokSize := len(token) * ratio
```

But string lengths are integer, so we have to use integers and Euclidean division.
If we were computing `tokSize = len(token) * totalOutputSize / totalInputSize` for all the tokens, the integer rounding errors would accumulate. And the sum of the length of the truncated tokens at the end wouldn’t be exactly `totalOutputSize`.
So, instead of that, we do a linear interpolation to place the boundaries of the tokens.

Here is what the variables of the algorithm are representing when we are computing the new size of the token at index 2:
![image](https://github.com/DataDog/test-infra-definitions/assets/1437785/db65c5cf-71a9-415e-a665-7366853efcac)

Reply:
Thanks for the explanation. Can you rename `prevY` to something like `prevOutputSize` and `X` to `totalTokenSize`?